### ISSUE TITLE
Linux_android flutter_gallery_sksl_warmup__transition_perf_e2e is 2.67% flaky

### ISSUE THREAD

**Issue by jmagman** [flutter/flutter#83298]
```
[flutter_gallery_sksl_warmup__transition_perf_e2e] [STDOUT] stdout: [        ] I/flutter ( 7731): > Text Fields@Cupertino
[flutter_gallery_sksl_warmup__transition_perf_e2e] [STDOUT] stdout: [+1172 ms] I/flutter ( 7731): The finder "zero widgets with text "Text Fields" (ignoring offstage widgets)" (used in a call to "tap()") could not find any matching widgets.
[flutter_gallery_sksl_warmup__transition_perf_e2e] [STDOUT] stdout: [   +2 ms] I/flutter ( 7731): #0      WidgetController._getElementPoint (package:flutter_test/src/controller.dart:902)
[flutter_gallery_sksl_warmup__transition_perf_e2e] [STDOUT] stdout: [        ] I/flutter ( 7731): #1      WidgetController.getCenter (package:flutter_test/src/controller.dart:841)
[flutter_gallery_sksl_warmup__transition_perf_e2e] [STDOUT] stdout: [        ] I/flutter ( 7731): #2      WidgetController.tap (package:flutter_test/src/controller.dart:273)
[flutter_gallery_sksl_warmup__transition_perf_e2e] [STDOUT] stdout: [        ] I/flutter ( 7731): #3      runDemos (file:///opt/s/w/ir/x/w/recipe_cleanup/tmpZWfox_/flutter%20sdk/dev/integration_tests/flutter_gallery/test_driver/run_demos.dart:65)
[flutter_gallery_sksl_warmup__transition_perf_e2e] [STDOUT] stdout: [        ] I/flutter ( 7731): <asynchronous suspension>
[flutter_gallery_sksl_warmup__transition_perf_e2e] [STDOUT] stdout: [        ] I/flutter ( 7731): <asynchronous suspension>
[flutter_gallery_sksl_warmup__transition_perf_e2e] [STDOUT] stdout: [        ] I/flutter ( 7731): (elided one frame from package:stack_trace)
[flutter_gallery_sksl_warmup__transition_perf_e2e] [STDOUT] stdout: [        ] I/flutter ( 7731): 01:48 +0 ~1: flutter gallery transitions on e2e all demos [E]
[flutter_gallery_sksl_warmup__transition_perf_e2e] [STDOUT] stdout: [        ] I/flutter ( 7731):   Test failed. See exception logs above.
[flutter_gallery_sksl_warmup__transition_perf_e2e] [STDOUT] stdout: [        ] I/flutter ( 7731):   The test description was: all demos
```
https://logs.chromium.org/logs/flutter/buildbucket/cr-buildbucket.appspot.com/8846488026204887728/+/u/run_flutter_gallery_sksl_warmup__transition_perf_e2e/stdout

Here's what it looks like it fails, when it's supposed to not find a `Text Fields` widget:

![drive_01](https://user-images.githubusercontent.com/682784/119405933-e51e0c00-bc96-11eb-86ce-8b90f4a76372.png)

2.67% flake rate.

**Comment by jmagman** [comment]
`engine` and assigning @zanderso per https://github.com/flutter/flutter/blob/de23eb7b292a5338d4133df986f8e8c4f5db8b6c/TESTOWNERS#L39.

**Comment by zanderso** [comment]
The same failure is happening on the non-sksl-warmup version of this benchmark: https://ci.chromium.org/ui/p/flutter/builders/prod/Linux%20flutter_gallery__transition_perf_e2e/1496/overview

**Comment by jmagman** [comment]
> The same failure is happening on the non-sksl-warmup version of this benchmark: https://ci.chromium.org/ui/p/flutter/builders/prod/Linux%20flutter_gallery__transition_perf_e2e/1496/overview

I was just about to disable. 😄

**Comment by zanderso** [comment]
Assigning @jason-simmons to try to get a local repro.

**Comment by keyonghan** [comment]
Adjusting the priority to P1 as the flakiness ratio exceeds SLO 2%.

**Comment by jason-simmons** [comment]
I tried running this benchmark script 15 times but only reproduced the `zero widgets with text "Text Fields"` error once.  The "Text Fields" list item was onscreen when the script failed.

Perhaps someone who knows the internals of Flutter Driver might have some insights about why this would fail?

**Comment by chinmaygarde** [comment]
During triage, we weren't certain if there was a non-racy way of finding widgets onscreen and the best the driver could do was to introduce conservatively long delays between the time the widgets are presented and the asserts run via the service protocol.

@dnfield @jonahwilliams Can you shed some light on how the Flutter driver runs its assertions? If this is inherently racy (which I suspect it is), we might need to rethink how these tests work.

**Comment by jonahwilliams** [comment]
The gist of this logic is:

https://github.com/flutter/flutter/blob/34059eec2c671472ee95f5093705fb9fed0b3919/dev/integration_tests/flutter_gallery/test_driver/transitions_perf_test.dart#L131-L148

Basically the test tries to scroll each item into view and then tap on it. It could fail if there is some disagreement on what constitutes visible versus tappable. Why that would be flaky I don't quite know, but since the app is just running normally it is possible that slight delays can lead to this being missed.


I could double check that at least the visible/tappable logic match up. If not, easy fix.

If they do already match up, then maybe we could introduce a retry (i.e. check if this is tappable, if not run scroll into view once more?)

**Comment by Hixie** [comment]
FWIW, we should definitely not use delays. There's no reason we can't synchronize here, we have a communication channel and can wait for the other side to say it is ready each time. (I believe that's what the code above is actually doing? Of course, why it's failing, I can't answer...)

**Comment by dnfield** [comment]
@jonahwilliams - your link is to a different version of this test.

I strongly suspect that https://github.com/flutter/flutter/blob/master/dev/integration_tests/flutter_gallery/test_driver/run_demos.dart#L70 is contributing to flakiness here.

**Comment by dnfield** [comment]
While that line is fishy, it probably isn't causing this flake since the demo before this one is not in the list of unsynchronized demos (e.g. it doesn't have an unbounded animation we have to wait on).

We may want to have a `controller.scrollUntilVisible(demoItem, 40.0)` after our `pageBack` logic every time. But it seems like the demo item is visible on the screen when it flakes, and we might just be lacking some synchronization needed before `controller.tap`

**Comment by chinmaygarde** [comment]
It looks like this is hard to reproduce. But even if there were a reliable way to reproduce this, the necessary tools to run such tests without races just doesn't don't exist. We may need to repurpose this issue to explore adding that synchrnoization mechanism for the tests (perhaps only in the debug runtime mode). In the meantime, should we disable this test?

**Comment by jonahwilliams** [comment]
I'm going to take a look at this today

**Comment by jonahwilliams** [comment]
actually I'm wrong, the error message above is printed if the finder does not evaluate to true anywhere. That would imply we're on the wrong page? or we scrolled offscreen

**Comment by jonahwilliams** [comment]
but maybe it got scrolled entirely offscreen?

**Comment by jonahwilliams** [comment]
does anyone know what the difference is between this and flutter_gallery_sksl_warmup__transition_perf? Do we have two tests that are 99% the same? If so lets save ourselves some time and delete this one

**Comment by zanderso** [comment]
I believe the e2e versions of benchmarks were intended to be less flaky due to having less reliance on flutter driver.

I'd be grateful if @CareF and/or @liyuqian could refresh our memories =)

**Comment by jonahwilliams** [comment]
it looks like this is the flaky one though ....

**Comment by CareF** [comment]
> I believe the e2e versions of benchmarks were intended to be less flaky due to having less reliance on flutter driver.
> 
> I'd be grateful if @CareF and/or @liyuqian could refresh our memories =)

The non-e2e version was previously more than 10% flaky before last summer, which I believe is mostly because of extra memory used for the logging that crashes the app and/or the timeout issue for flutter driver. The e2e version fixed this problem back then. The e2e version was also tested to have slightly better performance due to smaller test overhead. Otherwise, the two tests are intended as the same test case.

**Comment by zanderso** [comment]
Unless I'm misreading the infra metrics dashboard, neither has flaked in the past 15 days. @keyonghan, does the dashboard continue collecting stats for things that are marked flaky?

**Comment by keyonghan** [comment]
> Unless I'm misreading the infra metrics dashboard, neither has flaked in the past 15 days. @keyonghan, does the dashboard continue collecting stats for things that are marked flaky?

The flaky bug was filed based on the past 15 days data on that specific day. Here is the [dashboard](https://data.corp.google.com/sites/flutter_infra_metrics_datasite/flutter_luci_prod_build_flakiness_dashboard/?p=START_TIME:%222021-05-10%22&p=END_TIME:%222021-05-24%22&p=BUILDER_NAME:%22Mac_android%20microbenchmarks%22) for the past 15 day windows around 05/24 (the creation date of this bug).

**Comment by jmagman** [comment]
The last time I see a flake was 6/2/21:
https://ci.chromium.org/p/flutter/builders/prod/Mac_ios%20flutter_gallery__transition_perf_e2e_ios32/1124

```
[flutter_gallery__transition_perf_e2e_ios32] [STDOUT] stdout: flutter: > Text Fields@Cupertino
[flutter_gallery__transition_perf_e2e_ios32] [STDOUT] stdout: flutter: The finder "zero widgets with text "Text Fields" (ignoring offstage widgets)" (used in a call to "tap()") could not find any matching widgets.
```
https://logs.chromium.org/logs/flutter/buildbucket/cr-buildbucket.appspot.com/8845477934872445600/+/u/run_flutter_gallery__transition_perf_e2e_ios32/stdout

When it failed, the screen looked like this:
![drive_01-3](https://user-images.githubusercontent.com/682784/121233409-82f10980-c847-11eb-935a-50f51f27b24e.png)
https://storage.googleapis.com/flutter_logs/flutter/816a857a98237f948cbe4856a6d7d47fb48f6b71/flutter_gallery__transition_perf_e2e_ios32/ff9a7192-6676-4bb3-ab37-c2627a3ffbc8/drive_01.png

**Comment by jonahwilliams** [comment]
Text Fields is right there, I wonder if this is "just" a timing issue and the e2e test needs longer delays, because there is less "natural delay" from having an RPC layer between the test and host

**Comment by zanderso** [comment]
Copying from https://github.com/flutter/flutter/pull/84208#pullrequestreview-678868911 for visibility:

> In the logs for the flake, the exception is on line 65:
> 
> https://logs.chromium.org/logs/flutter/buildbucket/cr-buildbucket.appspot.com/8846488026204887728/+/u/run_flutter_gallery_sksl_warmup__transition_perf_e2e/stdout
> 
> Line 65 is in a for-loop that runs twice. From reading the code, my theory is that the tap on line 65 is failing on the second iteration due to either the pageBack() taking too long, or due to a need to do another scrollUntilVisible before trying another tap, or a combination of those two things.

**Comment by chinmaygarde** [comment]
My sense is that there is currently no way of doing this deterministically and hence the need for the gratuitous delays. @Hixie [seemed to suggest](https://github.com/flutter/flutter/issues/83298#issuecomment-852534630) we should build such a mechanism but that may be a longer term project that needs to be explicitly scoped out. @zanderso Not sure if this makes sense for the EngProd initiative. In the meantime, @jonahwilliams 's [patch bumps up the timeout](https://github.com/flutter/flutter/pull/84208#pullrequestreview-678868911) and perhaps we could land that to alleviate some of the "flakiness". Is that a reasonable summary of next steps?

**Comment by Hixie** [comment]
There's really not much to build. We already have the communication mechanism if we're using flutter driver. It's just a matter of sending a message from the host to the app and vice versa. That said, I don't really understand why there's any issue here. The driver commands are already synchronized. Do we have a clear understanding of what the actual race condition is?

**Comment by jonahwilliams** [comment]
this test is not using flutter_driver, its using integration_test, which is either not synchronized or synchronized differently

**Comment by Hixie** [comment]
Ah. My understanding was that integration_test is entirely in-app so there's nothing to synchronize. I may be mistaken, though.

**Comment by dnfield** [comment]
I think I know exactly what's wrong here.

**Comment by dnfield** [comment]
This is still happening after my change. :\

**Comment by zanderso** [comment]
@dnfield Not sure, but I wonder if this scroll back to the top when going to a new category is relevant: https://github.com/flutter/flutter/blob/master/dev/integration_tests/flutter_gallery/test_driver/run_demos.dart#L46. I also speculated here https://github.com/flutter/flutter/pull/84208#pullrequestreview-678868911 that a delay after the `pageBack` might be needed.

**Comment by chinmaygarde** [comment]
From triage, it looks like we still need more information here as recent attempts to deflake this have not been successful. This test is currently is also currently marked flaky on the dashboard so it should not be blocking us anymore. We are going to gather [more logs in the meantime](https://logs.chromium.org/logs/flutter/buildbucket/cr-buildbucket/8843864495448323057/+/u/run_flutter_gallery__transition_perf_e2e/stdout?format=raw). My confidence in how this mechanism is deterministic at all is lower than that of the room so I am going to pick someones mind about how this is working as well.

Dan is also going to work on dumping the widget tree when the exception happens.

**Comment by dnfield** [comment]
https://github.com/flutter/flutter/pull/84996 will dump the app and all elements the controller is aware of if the tap gesture fails.

**Comment by chinmaygarde** [comment]
@dnfield: Per @zanderso, we did not get the information we wanted because the app was running in profile mode? Is that correct? What are the next steps in debugging this?

**Comment by dnfield** [comment]
@chinmaygarde  - correct.

Someone needs to reproduce this locally and try to instrument the failure. I have a vague idea of how to do that. I'll give it a shot today or tomorrow.

**Comment by chinmaygarde** [comment]
Were you able to do a local repro for this? The dashboard does show that the flaky ratio is below the threshold now but it's a pretty small sample size.

**Comment by chinmaygarde** [comment]
We haven't been able to reproduce this and flakiness is below the threshold (1.25%). Additional logging was added but since this is profile mode, these are not fully available.

**Comment by fluttergithubbot** [comment]
Current flaky ratio for the past 15 days is 4.10%.
One recent flaky example for a same commit: https://ci.chromium.org/ui/p/flutter/builders/prod/Linux%20flutter_gallery_sksl_warmup__transition_perf_e2e/2089
Commit: https://github.com/flutter/flutter/commit/ff49d35c90cbe127620ebf797ab0862199650bc9
Flaky builds:
https://ci.chromium.org/ui/p/flutter/builders/prod/Linux%20flutter_gallery_sksl_warmup__transition_perf_e2e/2089
https://ci.chromium.org/ui/p/flutter/builders/prod/Linux%20flutter_gallery_sksl_warmup__transition_perf_e2e/2079
https://ci.chromium.org/ui/p/flutter/builders/prod/Linux%20flutter_gallery_sksl_warmup__transition_perf_e2e/2056
https://ci.chromium.org/ui/p/flutter/builders/prod/Linux%20flutter_gallery_sksl_warmup__transition_perf_e2e/2051
https://ci.chromium.org/ui/p/flutter/builders/prod/Linux%20flutter_gallery_sksl_warmup__transition_perf_e2e/2024

**Comment by keyonghan** [comment]
Reassigned to test owner for further triage, and also adjust the priority.
/cc @chunhtai shall we promote this to the auto process?

**Comment by chunhtai** [comment]
@keyonghan The script currently does not adjust the priority if priority has been manually adjusted. I can add that plus reassign if that makes more sense

**Comment by zanderso** [comment]
We are working on replacing this benchmark with a new one that will be less flaky: https://github.com/flutter/flutter/issues/86458. In the meantime, we will keep this benchmark running and marked flaky so that we can use the numbers that we are getting from successful runs to avoid performance regressions.

**Comment by fluttergithubbot** [comment]
Current flaky ratio for the past 15 days is 7.69%.
One recent flaky example for a same commit: https://ci.chromium.org/ui/p/flutter/builders/prod/Linux%20flutter_gallery_sksl_warmup__transition_perf_e2e/2147
Commit: https://github.com/flutter/flutter/commit/9c32e5a1d2e4a59dc67494755ff7c4881b9d5e50
Flaky builds:
https://ci.chromium.org/ui/p/flutter/builders/prod/Linux%20flutter_gallery_sksl_warmup__transition_perf_e2e/2147
https://ci.chromium.org/ui/p/flutter/builders/prod/Linux%20flutter_gallery_sksl_warmup__transition_perf_e2e/2140
https://ci.chromium.org/ui/p/flutter/builders/prod/Linux%20flutter_gallery_sksl_warmup__transition_perf_e2e/2116
https://ci.chromium.org/ui/p/flutter/builders/prod/Linux%20flutter_gallery_sksl_warmup__transition_perf_e2e/2112
https://ci.chromium.org/ui/p/flutter/builders/prod/Linux%20flutter_gallery_sksl_warmup__transition_perf_e2e/2106
https://ci.chromium.org/ui/p/flutter/builders/prod/Linux%20flutter_gallery_sksl_warmup__transition_perf_e2e/2089
https://ci.chromium.org/ui/p/flutter/builders/prod/Linux%20flutter_gallery_sksl_warmup__transition_perf_e2e/2079
https://ci.chromium.org/ui/p/flutter/builders/prod/Linux%20flutter_gallery_sksl_warmup__transition_perf_e2e/2056
https://ci.chromium.org/ui/p/flutter/builders/prod/Linux%20flutter_gallery_sksl_warmup__transition_perf_e2e/2051

**Comment by chinmaygarde** [comment]
I have setup a meeting next week to discuss what we should do about such tests as we have typically done a poor job of measuring the underlying metrics in a non-flaky manner.

**Comment by fluttergithubbot** [comment]
Current flaky ratio for the past 15 days is 10.00%.
One recent flaky example for a same commit: https://ci.chromium.org/ui/p/flutter/builders/prod/Linux%20flutter_gallery_sksl_warmup__transition_perf_e2e/2147
Commit: https://github.com/flutter/flutter/commit/9c32e5a1d2e4a59dc67494755ff7c4881b9d5e50
Flaky builds:
https://ci.chromium.org/ui/p/flutter/builders/prod/Linux%20flutter_gallery_sksl_warmup__transition_perf_e2e/2147
https://ci.chromium.org/ui/p/flutter/builders/prod/Linux%20flutter_gallery_sksl_warmup__transition_perf_e2e/2140
https://ci.chromium.org/ui/p/flutter/builders/prod/Linux%20flutter_gallery_sksl_warmup__transition_perf_e2e/2116
https://ci.chromium.org/ui/p/flutter/builders/prod/Linux%20flutter_gallery_sksl_warmup__transition_perf_e2e/2112
https://ci.chromium.org/ui/p/flutter/builders/prod/Linux%20flutter_gallery_sksl_warmup__transition_perf_e2e/2106
https://ci.chromium.org/ui/p/flutter/builders/prod/Linux%20flutter_gallery_sksl_warmup__transition_perf_e2e/2089

**Comment by fluttergithubbot** [comment]
Current flaky ratio for the past 15 days is 5.26%.
One recent flaky example for a same commit: https://ci.chromium.org/ui/p/flutter/builders/prod/Linux_android%20flutter_gallery_sksl_warmup__transition_perf_e2e/94
Commit: https://github.com/flutter/flutter/commit/16b5af9bb1154c26e36c4aa5c09f07646c7f32d1
Flaky builds:
https://ci.chromium.org/ui/p/flutter/builders/prod/Linux_android%20flutter_gallery_sksl_warmup__transition_perf_e2e/94
https://ci.chromium.org/ui/p/flutter/builders/prod/Linux_android%20flutter_gallery_sksl_warmup__transition_perf_e2e/56
https://ci.chromium.org/ui/p/flutter/builders/prod/Linux_android%20flutter_gallery_sksl_warmup__transition_perf_e2e/30
https://ci.chromium.org/ui/p/flutter/builders/prod/Linux_android%20flutter_gallery_sksl_warmup__transition_perf_e2e/27
https://ci.chromium.org/ui/p/flutter/builders/prod/Linux_android%20flutter_gallery_sksl_warmup__transition_perf_e2e/18

**Comment by chunhtai** [comment]
Even for a passing builds, it can still be flaky due to test runner auto rerun, see https://logs.chromium.org/logs/flutter/buildbucket/cr-buildbucket.appspot.com/8839872728933611088/+/u/check_flaky/stdout

**Comment by zanderso** [comment]
@keyonghan @godofredoc We'd like to move the `flutter_gallery...transition_perf` family of benchmarks out of the prod pool, and run them on staging. But before we do that, we wanted to verify with you that these benchmarks would still run and send results to SkiaPerf, and that there is still a dashboard where we could view their status.

**Comment by keyonghan** [comment]
Existing staging tasks are not sending results to SkiaPerf as of now, but we can set exception for these tests.
Created https://github.com/flutter/flutter/issues/88296 to track.

**Comment by chinmaygarde** [comment]
Closing in favor of https://github.com/flutter/flutter/issues/88296
